src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -122,145 +493,157 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
-				bufPtr = goAndCopyNextWord(smooth, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
+				bufPtr = goAndCopyNextWord(tmpbuf, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
 #ifdef _IRR_DEBUG_OBJ_LOADER_
-	os::Printer::log("Loaded smoothing group start",smooth, ELL_DEBUG);
+	os::Printer::log("Loaded smoothing group start",tmpbuf, ELL_DEBUG);
 #endif
-				if (core::stringc("off")==smooth)
-					smoothingGroup=0;
+				if (strcmp("off", tmpbuf)==0)
+					smoothingGroup=0u;
 				else
-                    sscanf(smooth,"%u",&smoothingGroup);
+                    sscanf(tmpbuf,"%u",&smoothingGroup);
 			}
 			break;
 
 		case 'u': // usemtl
 			// get name of material
 			{
-				char matName[WORD_BUFFER_LENGTH];
-				bufPtr = goAndCopyNextWord(matName, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
+				bufPtr = goAndCopyNextWord(tmpbuf, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
 #ifdef _IRR_DEBUG_OBJ_LOADER_
-	os::Printer::log("Loaded material start",matName, ELL_DEBUG);
+	os::Printer::log("Loaded material start",tmpbuf, ELL_DEBUG);
 #endif
-				mtlName=matName;
+				mtlName=tmpbuf;
 
                 if (ctx.useMaterials && !ctx.useGroups)
                 {
                     asset::IAsset::E_TYPE types[] {asset::IAsset::ET_SUB_MESH, (asset::IAsset::E_TYPE)0u };
-                    auto mb_bundle = _override->findCachedAsset(genKeyForMeshBuf(ctx, _file->getFileName().c_str(), mtlName, grpName), types, ctx.inner, 1u).getContents();
-                    if (mb_bundle.first!=mb_bundle.second)
+                    auto mb_bundle = _override->findCachedAsset(genKeyForMeshBuf(ctx, _file->getFileName().c_str(), mtlName, grpName), types, ctx.inner, _hierarchyLevel+ICPUMesh::MESHBUFFER_HIERARCHYLEVELS_BELOW);
+                    auto mbs = mb_bundle.getContents();
+                    if (mbs.first!= mbs.second)
                     {
-                        auto mb = static_cast<asset::ICPUMeshBuffer*>(mb_bundle.first->get());
-                        mb->grab();
-                        SObjMtl* mtl = findMtl(ctx, mtlName, grpName);
-                        ctx.preloadedSubmeshes.insert(std::make_pair(mtl, mb));
+                        submeshes.push_back(core::smart_refctd_ptr_static_cast<ICPUMeshBuffer>(*mbs.first));
                     }
-                    else mtlChanged=true;
-
-                    submeshLoadedFromCache = (mb_bundle.first!=mb_bundle.second);
+                    else
+                    {
+                        submeshes.push_back(core::make_smart_refctd_ptr<ICPUMeshBuffer>());
+                        auto found = pipelines.find(mtlName);
+                        if (found != pipelines.end())
+                        {
+                            auto& pipeln = found->second;
+                            //cloning pipeline because it will be edited (vertex input params, shaders, ...)
+                            //note shallow copy (depth=0), i.e. only pipeline is cloned, but all its sub-assets are taken from original object
```

Comment:
even though you clone, you should cache it in the MTL loader (so the total number of clones is limited)

Reply:
>(so the total number of clones is limited)

? i dont understand

Reply:
there's only a certain (given) max number of permutations of such a pipeline for all possible vertex inputs, blend states, and shaders attached... shouldn't be cloning a new one, if you've already made an identical one.